core/sql/ustat/hs_globals.cpp

```diff
@@ -994,7 +994,8 @@ void HSGlobalsClass::formGroupSets()
       {
          if (LM->LogNeeded())
          {
-            sprintf(LM->msg, "\tMC: GROUP (%s) has state DONT_TRY, is skipped", mgroup_set->colNames->data());
+            sprintf(LM->msg, "\tMC: GROUP (%s) has state DONT_TRY, is skipped", 
+              LM->truncate(mgroup_set->colNames->data(),sizeof(LM->msg-200)));
```

Comment:
I wonder if there is a slightly better way to do this so that the total msg length is exactly 2000. 

One way to do so would be to compute the the skeleton string length L (such as strlen("\MC: GROUP () has state DONT_TRY, is skipped"), and then set the truncation length to 2000 - L. 

When L is available, we can supply it to sprint() as follows. See https://stackoverflow.com/questions/2239519/is-there-a-way-to-specify-how-many-characters-of-a-string-to-print-out-using-pri.

printf ("Here are the first %d chars: %.*s\n", L, L, "A string that is more than 8 chars");

Reply:
It may also be a good idea to use snprintf().

Reply:
This is complicated in general. For example, some of the log messages have multiple parameters substituted in the sprintf. I would need to do several sprintfs, each one calculating the L from the previous substitution. There are many examples of this. Too, where two strings are involved that may need truncation, I would need additional logic to evenly apportion the available space to them. 

My feeling is the effort does not justify the benefit.

I did consider using snprintf. The disadvantage there is that it would truncate the end of the message. That might leave out interesting and vital information that today follows the column name list. Instead, I truncated the column name list and visually checked in all cases that I truncated it enough that the end of the message fits within the sprintf buffer.

I don't claim this is ideal. A better approach would be to completely rewrite the HSLogMan methods to use a C++ streaming style into, say, an NAString, eliminating the need for sprintf. I could avoid truncation completely, having HSLogMan fold the message into multiple lines if the underlying logger has any line length limitations. However, the code changes would be truly massive; logging is ubiquitous throughout the UPDATE STATISTICS code. Again, I did not feel that the effort was justified.

Reply:
Actually, let me think about the HSLogMan refactoring. I could provide new methods alongside the old, and then convert only the problem code paths to use the new methods. I will consider that.